### PR TITLE
[Distroless] Convert the GCE manifests for master containers.

### PR DESCRIPTION
**Pull request by yuwenma**
* Touched containers: kube-apiserver, kube-scheduler,
kube-controller-manager.
* Remove the shell dependencies when upstart the containers.
* Reformat the command parameters to ["Exec", "Param1", "Param2"]
* Remove log direction, using flags to export logs to the expected path.   

Test: The manifest change is tested on a real GKE cluster (GKE sandbox). Here's the [Master VM](https://pantheon.corp.google.com/compute/instancesDetail/zones/us-central1-c/instances/gke-ae6a5d2c1b849bb2cfd1-4b84-6703-vm?project=gke-yuwenma-hosted-master&organizationId=433637338589)

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug

/kind cleanup

> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR is a prerequisite for switching the core master containers' base image to distroless/static.(See [this kep](https://github.com/kubernetes/enhancements/pull/900) for the reasons why we want to change to distroless/static).
Basically it removes all the shell dependencies (log redirection, using shell to kick-off a kube binary file) which isn't supported by distroless.  

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note None

```

### CODE REVIEW

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -2059,7 +2079,7 @@ function start-kube-controller-manager {
     params+=" --concurrent-service-syncs=${CONCURRENT_SERVICE_SYNCS}"
   fi
   if [[ "${NETWORK_PROVIDER:-}" == "kubenet" ]]; then
-    params+=" --allocate-node-cidrs=true"
```
Comment (yuwenma) [inline cluster/gce/gci/configure-helper.sh]: Specifying the `true` is not acceptable. (Will cause kube-controller-manager container unable to start).

**Comment by yuwenma** [comment]
/assign @tallclair 
/assign @MaciekPytel  
  
May I get a  /ok-to-test label?

**Comment by YuZongYangHi** [comment]
hahahaa

**Comment by yuwenma** [comment]
Gentle ping, can I get a review on this @MaciekPytel @tallclair :)

File: cluster/gce/gci/apiserver_manifest_test.go [hunk 2]
```diff
@@ -27,7 +27,7 @@ import (
 	"strings"
 	"testing"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (tallclair) [inline cluster/gce/gci/apiserver_manifest_test.go]: nit: revert this (unnecessary)

btw - I've been seeing this change a lot lately, is there some tool that does this automatically?

File: cluster/gce/manifests/kube-apiserver.manifest [hunk 3]
```diff
@@ -25,10 +25,10 @@
       }
     },
     "command": [
-                 "/bin/sh",
-                 "-c",
-                 "exec /usr/local/bin/kube-apiserver {{params}} --allow-privileged={{pillar['allow_privileged']}} 1>>/var/log/kube-apiserver.log 2>&1"
-               ],
+      "/usr/local/bin/kube-apiserver",
+      "--allow-privileged={{pillar['allow_privileged']}}",
```
Comment (tallclair) [inline cluster/gce/manifests/kube-apiserver.manifest]: Can we separate command & args? That will also simplify the test.

File: cluster/gce/manifests/kube-controller-manager.manifest [hunk 4]
```diff
@@ -25,10 +25,9 @@
       }
     },
     "command": [
-                 "/bin/sh",
-                 "-c",
-                 "exec /usr/local/bin/kube-controller-manager {{params}} 1>>/var/log/kube-controller-manager.log 2>&1"
-               ],
+      "/usr/local/bin/kube-controller-manager",
+      {{params}}
```
Comment (tallclair) [inline cluster/gce/manifests/kube-controller-manager.manifest]: ditto re: separate args

same below.

File: cluster/gce/manifests/kube-apiserver.manifest (hunk 3 above)
Comment (tallclair) [reply cluster/gce/manifests/kube-apiserver.manifest]: To clarify, I'm proposing:

```
"command": ["/usr/local/bin/kube-apiserver"],
"args": [
  "--allow-privileged={{pillar['allow_privileged']}}",
  {{params}}
],
```

File: cluster/gce/gci/apiserver_manifest_test.go (hunk 2 above)
Comment (yuwenma) [reply cluster/gce/gci/apiserver_manifest_test.go]: Yeah. My IDE has a google3 code format plugin, which I think automatically updates the imports to the internal format.

**Review by yuwenma — COMMENTED** [review]
Updated.

File: cluster/gce/manifests/kube-apiserver.manifest [hunk 5]
```diff
@@ -25,10 +25,12 @@
...
-                 "/bin/sh",
-                 "-c",
-                 "exec /usr/local/bin/kube-apiserver {{params}} --allow-privileged={{pillar['allow_privileged']}} 1>>/var/log/kube-apiserver.log 2>&1"
-               ],
+    	"/usr/local/bin/kube-apiserver"
+    ],
+	"args": [
+  		"--allow-privileged={{pillar['allow_privileged']}}",
+  		{{params}}
+	],
```
Comment (tallclair) [inline cluster/gce/manifests/kube-apiserver.manifest]: nit: fix indentation. Looks like a tabs vs. spaces thing. Same in other manifests.

File: cluster/gce/gci/configure-helper.sh [hunk 6]
```diff
@@ -2090,9 +2110,13 @@ function start-kube-controller-manager {
...
   if [[ -n "${RUN_CONTROLLERS:-}" ]]; then
-    params+=" --controllers=${RUN_CONTROLLERS}"
+    # Trim the `RUN_CONTROLLERS` value. This field is quoted which is
+    # incompatible with the `convert-manifest-params` format.
+    params+=" --controllers=${RUN_CONTROLLERS//\'}"
   fi
-
+  params+=" --log-file=${LOG_PATH}"
+  params+=" --logtostderr=false"
+  params=$(convert-manifest-params "${params}")
```
Comment (tallclair) [inline cluster/gce/gci/configure-helper.sh]: I think `$(...)` should be quoted. I'm not sure it matters in this specific case, but it's fragile without the quotes, especially since it includes newline characters.
```suggestion
  params="$(convert-manifest-params "${params}")"
```

Same applies elsewhere.

File: cluster/gce/gci/configure-helper.sh [hunk 7]
```diff
@@ -25,6 +25,21 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+function convert-manifest-params {
+  # Convert the manifest's command's parameters.
+  # Previously the command is in the form of
+  # ["/bin/sh", "-c", "exec KUBE_EXEC_BINARY --param1=val1 --param2-val2"].
+  # Since we remove the bash dependency and the executable binary is called
+  # directly, the format should be changed to
+  # ["KUBE_EXEC_BINARY", "--param1=val1", "--param2-val2"].
```
Comment (tallclair) [inline cluster/gce/gci/configure-helper.sh]: I think this comment is no longer accurate?

File: cluster/gce/gci/configure-helper.sh [hunk 8]
```diff
@@ -25,6 +25,21 @@ set -o errexit
...
+  IFS=' ' read -ra FLAGS <<< "$1"
+  params=""
+  for flag in "${FLAGS[@]}"; do
+    params+="\n\"$flag\","
+  done
+  echo "${params::-1}"
```
Comment (tallclair) [inline cluster/gce/gci/configure-helper.sh]:
```suggestion
  echo "${params::-1}"  # drop trailing comma
```

This will be an error if FLAGS is ever empty - but maybe that should never happen?

Comment (yuwenma) [reply cluster/gce/gci/configure-helper.sh]: I was thinking of that, but should not happen since this is dedicated to GCP kcm/apiserver/scheduler.

File: cluster/gce/gci/configure-helper.sh (hunk 7 above)
Comment (yuwenma) [reply cluster/gce/gci/configure-helper.sh]: haha, good catch!

**Comment by MaciekPytel** [comment]
/approve
Looks fine, but I don't have enough context on this change to feel comfortable giving lgtm. I'll leave that to @tallclair.

**Comment by dims** [comment]
Looks like this is all set :)

**Comment by tallclair** [comment]
> Looks like this is all set :)

I just wanted to show off what your klog changes are enabling :)

**Comment by dims** [comment]
Thanks!!!!

**Comment by wojtek-t** [comment]
We need to revert that actually - this broke scalability tests in a way that we have only last 1.8GB of logs (i.e. it doesn't work with logrotate). We are blindless without logs.
Revert in #76190
